pkg/services/accesscontrol/resourcepermissions/service.go

```diff
@@ -0,0 +1,226 @@
+}
+
+func (s *Service) GetPermissions(ctx context.Context, orgID int64, resourceID string) ([]accesscontrol.ResourcePermission, error) {
+	return s.store.GetResourcesPermissions(ctx, orgID, accesscontrol.GetResourcesPermissionsQuery{
+		Actions:     s.actions,
+		Resource:    s.options.Resource,
+		ResourceIDs: []string{resourceID},
+	})
+}
+
+func (s *Service) SetUserPermission(ctx context.Context, orgID, userID int64, resourceID string, actions []string) (*accesscontrol.ResourcePermission, error) {
+	if !s.options.Assignments.Teams {
+		return nil, ErrInvalidAssignment
+	}
+
+	if !s.validateActions(actions) {
+		return nil, ErrInvalidActions
+	}
+
+	if err := s.validateResource(ctx, orgID, resourceID); err != nil {
+		return nil, err
+	}
+
+	if err := s.validateUser(ctx, orgID, userID); err != nil {
+		return nil, err
+	}
+
+	return s.store.SetUserResourcePermission(ctx, orgID, userID, accesscontrol.SetResourcePermissionCommand{
+		Actions:    actions,
+		ResourceID: resourceID,
+		Resource:   s.options.Resource,
+	})
+}
+
+func (s *Service) SetTeamPermission(ctx context.Context, orgID, teamID int64, resourceID string, actions []string) (*accesscontrol.ResourcePermission, error) {
+	if !s.options.Assignments.Teams {
+		return nil, ErrInvalidAssignment
+	}
+	if !s.validateActions(actions) {
+		return nil, ErrInvalidActions
+	}
+
+	if err := s.validateTeam(ctx, orgID, teamID); err != nil {
+		return nil, err
+	}
+
+	if err := s.validateResource(ctx, orgID, resourceID); err != nil {
+		return nil, err
+	}
+
+	return s.store.SetTeamResourcePermission(ctx, orgID, teamID, accesscontrol.SetResourcePermissionCommand{
+		Actions:    actions,
+		ResourceID: resourceID,
+		Resource:   s.options.Resource,
+	})
+}
+
+func (s *Service) SetBuiltInRolePermission(ctx context.Context, orgID int64, builtInRole string, resourceID string, actions []string) (*accesscontrol.ResourcePermission, error) {
+	if !s.options.Assignments.BuiltInRoles {
+		return nil, ErrInvalidAssignment
+	}
+
+	if !s.validateActions(actions) {
+		return nil, ErrInvalidActions
+	}
+
+	if err := s.validateBuiltinRole(ctx, builtInRole); err != nil {
+		return nil, err
+	}
+
+	if err := s.validateResource(ctx, orgID, resourceID); err != nil {
+		return nil, err
+	}
+
+	return s.store.SetBuiltInResourcePermission(ctx, orgID, builtInRole, accesscontrol.SetResourcePermissionCommand{
+		Actions:    actions,
+		ResourceID: resourceID,
+		Resource:   s.options.Resource,
+	})
+}
+
+func (s *Service) MapActions(permission accesscontrol.ResourcePermission) (string, bool) {
+	for _, p := range s.permissions {
+		if permission.Contains(s.options.PermissionsToActions[p]) {
+			return p, true
+		}
+	}
+	return "", false
+}
+
+func (s *Service) MapPermission(permission string) []string {
+	for k, v := range s.options.PermissionsToActions {
+		if permission == k {
+			return v
+		}
+	}
+	return []string{}
+}
+
+func (s *Service) validateResource(ctx context.Context, orgID int64, resourceID string) error {
```

Comment:
Out of curiosity what is the use case for this? Prevent setting permissions on specific resources (like the default ones maybe?)

Reply:
Mostly ensuring that a resource actually exists. But could be used for other things